### ISSUE TITLE
as.class throws "Variable and function name style should be snake_case."

### ISSUE THREAD

**Issue by msberends** [jimhester/lintr#422]
This is a false positive: I define a new S3 class `mic` in my package for **M**inimal **I**nhibitory **C**oncentrations in antibiotics. They have S3 methods for `plot`, `summary`, you name it.

But my `as.mic()` and `is.mic()` function names throws this note:
```
Variable and function name style should be snake_case.
```

That's of course not true for these ones.

Great package btw, really helps keeping everything clean!!

**Comment by msberends** [comment]
Also the case for using parameter `na.rm`. I mean... common, you can't take that one from R lovers :)

Maybe an internal vector of exclusions would be wise here. If they follow R meticulously (like `as.class` and `na.rm`), they should be allowed and not throw a note.

**Comment by russHyde** [comment]
From the description, I don't think `is.mic` is a method on `mic`.
If you try running `methods(class = "mic")` you should get a list of the methods on `mic`.
If it were a method on `mic`, you wouldn't be able to pass it an object that wasn't an `mic`, and so it's output would always be TRUE.
By similar reasoning, `as.mic` shouldn't be a method on `mic`, since in that case it could only be passed a `mic`, which is already a `mic`.
When calling them, you would use `as.mic(some_object)`, whereas for your `mic`-methods, you'd call them using `plot(some_mic)`. So, lintr is correctly calling a snake-case issue on your newly defined functions (you could modify your config to allow dotted-case object names, but it might be more consistent with the rest of your code to use `is_mic` and `as_mic`)

Could you expand on the `na.rm` thing in a separate issue, and leave a reproducible example of what's happening

**Comment by msberends** [comment]
> From the description, I don't think `is.mic` is a method on `mic`.

I don't get want you mean here. Of course it's not a method, it's a *test*, consistent with any `is.someclass()` in R:

```r
> is.data.frame
#> function (x) 
#> inherits(x, "data.frame")
```

It just tests if parameter `x` is of class `someclass`.

> By similar reasoning, `as.mic` shouldn't be a method on `mic`, since in that case it could only be passed a `mic`, which is already a `mic`.

Exactly, and that's why `as.mic` is indeed not a method on `mic` at all, and I don't know where you got that from. It takes any input and tries to coerce it to a new class `mic`. Exactly the same logic as e.g. `as.Date()` and `as.numeric()`. So if I run lintr on base R, it would throw the same note as the one I got, for all these `as.xxx` functions. Now you tell me if that's expected behaviour or if it's false positives. I'm guessing the latter.

> but it might be more consistent with the rest of your code to use `is_mic` and `as_mic`

I very strongly disagree here. Using `as.someclass()` to coerce values to a new class `someclass` is exactly how (base) R does this (think of `as.character()`, `as.Date()`, `as.data.frame()`, ...) and I follow this compliance. I know that tidyverse does not follow this, which is frankly one of the very few things I dislike and don't understand about tidyverse (`as_tibble()` should really be `as.tbl_df()` to follow the R structure), but that's another discussion. 

For `na.rm`, I will file a new issue.

**Comment by russHyde** [comment]
I must have misunderstood something.

`is.mic` throwing a snake-case lint isn't a false positive by lintr's specification: it's a newly defined function that uses dotted.case. If you want lintr to bypass this function, you could either admit dotted-case objects (by modifying your config), rename this function, or wrap its definition in `# nolint start` `# nolint end` flags.

The reason `plot.mic` (etc) don't throw snake-case lints is because lintr can work out that they are S3 methods on `mic` - the dot preceding `mic` being a necessary part of the syntax and something that lintr is aware of.

**Comment by msberends** [comment]
I see. Maybe then it's an idea that the linter checks if the `x` part in `is.x()` is a defined S3 class in the package. Because when I would review a package with a newly defined S3 class, I would certainly expect the class definition to be accompanied by a `is.x()` function to check if the input inherits the new class.

**Comment by MichaelChirico** [comment]
My first instinct is to close this, however, the tidyverse itself is inconsistent on using `is_` (which would be suggested by the style guide), and often resorts to `is.`:

```
is.src, is.tbl, is_grouped_df, is.grouped_df, is_vector, is_bare_logical, is_scalar_logical, is_empty, is_scalar_integer, is_bare_integer, is_numeric, is_bare_vector, is_rate, is_scalar_vector, is_character, is_formula, is_bare_numeric, is_integer, is_logical, is_list, is_atomic, is_double, is_null, is_bare_list, is_bare_double, is_function, is_scalar_double, is_scalar_list, is_scalar_atomic, is_bare_character, is_scalar_character, is_bare_atomic, is_scalar_numeric, is.tibble, is_tibble, is.ggproto, is.Coord, is.ggplot, is.facet, is.theme
```

^ (from matching `^is` to all `getNamespaceExports` of `tidyverse` packages)

I still lean towards closing since it does appear tidyverse tries to offer `is_` equivalents in most cases, meaning the tidyverse recommendation would be to do the same and then `# nolint` the `is.` version. WDYT @AshesITR ?

**Comment by russHyde** [comment]
Would it be reasonable to introduce an 'allowed_prefixes' for use in the object-name linter. I think I'd struggle to neatly implement a custom linter that allows is.MyClass for multiple newly defined classes.

**Comment by AshesITR** [comment]
IINM the lint would go away if the package imports `as` or `is` from `{methods}` respectively.
cf. https://github.com/jimhester/lintr/blob/00ec31061f036f0c7ec5daa804cb9edb545fc5f8/R/object_name_linters.R#L57

**Comment by msberends** [comment]
Then I think we can close this without having a solid and R compliant solution. Importing `as` and `is` from methods is to me a bit weird and overkill. Also not advised in the R developers guidelines. I stand with my comments above and also agree that the tidyverse is inconsequent to this end. 

I personally think `as_` and `is_` are too colloquial. The formal R way is just `as.` and `is.` as I’ve pointed out above and how we know it from the ‘vanilla’ R language. Ergo, they should not get a note from lintr :p

Anyway, apparently no way out here, just a difference in opinion and perspective. And that’s okay!
